### PR TITLE
plan, executor: avoid children type assertion.

### PR DESCRIPTION
**Pull request by hanfei1991**
store LogicalPlan and PhysicalPlan directly, avoid type assertion.

### CODE REVIEW

**Comment by hanfei1991** [comment]
After Optimize

BenchmarkOptimize/select_count(*)_from_t_group_by_e-4                              30000             49761 ns/op           37488 B/op        494 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_10_group_by_e-4                10000            106967 ns/op           75809 B/op        969 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_50-4                           10000            110881 ns/op           76041 B/op        971 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_c_>_'1'_group_by_b-4                10000            146000 ns/op           93537 B/op       1301 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_=_1_group_by_b-4                  10000            127724 ns/op           87361 B/op       1154 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_>_1_group_by_b-4                  10000            133188 ns/op           90513 B/op       1172 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_20-4                         10000            146397 ns/op           99327 B/op       1272 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_30-4                         10000            151664 ns/op           99328 B/op       1272 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_40-4                         10000            151845 ns/op           99326 B/op       1272 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_50-4                         10000            150145 ns/op           99326 B/op       1272 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_40-4                                30000             58253 ns/op           39181 B/op        472 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_50-4                                30000             61256 ns/op           39182 B/op        472 allocs/op
BenchmarkOptimize/select_*_from_t_where_1_and_t.b_<=_50-4                          20000             59723 ns/op           39358 B/op        477 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_100_order_by_t.a_limit_1-4          10000            208626 ns/op          150961 B/op       1773 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_1_order_by_t.a_limit_10-4           10000            209091 ns/op          149480 B/op       1736 allocs/op
BenchmarkOptimize/select_*_from_t_use_index(b)_where_b_=_1_order_by_a-4            50000             38242 ns/op           23946 B/op        325 allocs/op
BenchmarkOptimize/select_*_from_t_where_d_<_cast('1991-09-05'_as_datetime)-4       20000             66599 ns/op           38864 B/op        492 allocs/op
BenchmarkOptimize/select_*_from_t_where_ts_<_'1991-09-05'-4                        20000             67803 ns/op           38768 B/op        490 allocs/op

Before Optimize

BenchmarkOptimize/select_count(*)_from_t_group_by_e-4                              30000             52170 ns/op           37488 B/op        494 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_10_group_by_e-4                10000            113133 ns/op           75809 B/op        969 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_<=_50-4                           10000            111260 ns/op           76041 B/op        971 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_c_>_'1'_group_by_b-4                10000            174081 ns/op           93537 B/op       1301 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_=_1_group_by_b-4                  10000            138265 ns/op           87361 B/op       1154 allocs/op
BenchmarkOptimize/select_count(*)_from_t_where_e_>_1_group_by_b-4                  10000            142825 ns/op           90513 B/op       1172 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_20-4                         10000            153970 ns/op           99309 B/op       1271 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_30-4                         10000            153404 ns/op           99309 B/op       1271 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_40-4                         10000            150491 ns/op           99310 B/op       1271 allocs/op
BenchmarkOptimize/select_count(e)_from_t_where_t.b_<=_50-4                         10000            156228 ns/op           99310 B/op       1271 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_40-4                                30000             57826 ns/op           39172 B/op        472 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_50-4                                30000             58881 ns/op           39172 B/op        472 allocs/op
BenchmarkOptimize/select_*_from_t_where_1_and_t.b_<=_50-4                          20000             60819 ns/op           39348 B/op        477 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_100_order_by_t.a_limit_1-4           5000            209446 ns/op          150961 B/op       1773 allocs/op
BenchmarkOptimize/select_*_from_t_where_t.b_<=_1_order_by_t.a_limit_10-4            5000            209065 ns/op          149441 B/op       1735 allocs/op
BenchmarkOptimize/select_*_from_t_use_index(b)_where_b_=_1_order_by_a-4            50000             39165 ns/op           23932 B/op        324 allocs/op
BenchmarkOptimize/select_*_from_t_where_d_<_cast('1991-09-05'_as_datetime)-4       20000             66688 ns/op           38864 B/op        492 allocs/op
BenchmarkOptimize/select_*_from_t_where_ts_<_'1991-09-05'-4                        20000             69144 ns/op           38768 B/op        490 allocs/op

File: executor/builder.go [hunk 1]
```diff
@@ -995,7 +995,6 @@ func (b *executorBuilder) buildAnalyzeColumnsPushdown(task plan.AnalyzeColumnsTa
 func (b *executorBuilder) buildAnalyze(v *plan.Analyze) Executor {
 	e := &AnalyzeExec{
 		baseExecutor: newBaseExecutor(v.Schema(), b.ctx),
-		tasks:        make([]*analyzeTask, 0, len(v.Children())),
```
Comment (zimulala) [inline executor/builder.go]: Why remove this? We can replace it with `len(v.ColTasks)+len(v.IdxTasks)`.

File: plan/planbuilder.go [hunk 2]
```diff
@@ -945,6 +946,7 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
 	if b.err != nil {
 		return nil
 	}
+	insertPlan.ResolveIndices()
```
Comment (zimulala) [inline plan/planbuilder.go]: Why remove it to here?

Comment (winoros) [reply plan/planbuilder.go]: We do `ResolveIndices` after building physical plan.

Comment (hanfei1991) [reply plan/planbuilder.go]: I moved `ResolveIndices` to interface `PhysicalPlan`, so `Insert`/`Update`/`Delete` can call resolveIndices individually.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**